File: qstack/qunimbus/core/governance.py
"""Deterministic governance and voting primitives.

This module provides decentralized governance for the QRATUM ghost machine,
enabling protocol changes through on-chain voting with stake-weighted decisions,
time-locked execution, and Merkle-logged votes.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Optional, Any
from enum import Enum


class VoteDecision(Enum):
    """Vote decision options."""
    APPROVE = "approve"
    REJECT = "reject"
    ABSTAIN = "abstain"


class ProposalStatus(Enum):
    """Proposal lifecycle status."""
    ACTIVE = "active"
    APPROVED = "approved"
    REJECTED = "rejected"
    EXECUTED = "executed"
    VETOED = "vetoed"


@dataclass(frozen=True)
class PolicyProposal:
    """A governance policy proposal.
    
    Attributes:
        id: Unique proposal identifier (hex string)
        title: Human-readable proposal title
        description: Detailed proposal description
        payload: Implementation-specific proposal data
        proposer: Validator ID who proposed (hex string)
        threshold: Required approval percentage (0-100)
        voting_period: Number of epochs for voting
        timelock: Number of epochs after approval before execution
        creation_epoch: Epoch when proposal was created
    """
    id: str
    title: str
    description: str
    payload: bytes
    proposer: str
    threshold: int
    voting_period: int
    timelock: int
    creation_epoch: int


@dataclass
class Vote:
    """A vote on a proposal.
    
    Attributes:
        voter: Validator ID (hex string)
        proposal_id: Proposal being voted on (hex string)
        decision: Vote decision
        weight: Voting weight (typically stake amount)
        signature: Vote signature (hex string)
        epoch: Epoch when vote was cast
    """
    voter: str
    proposal_id: str
    decision: VoteDecision
    weight: int
    signature: str
    epoch: int


@dataclass
class Authority:
    """An authority with veto power.
    
    Attributes:
        id: Authority identifier (hex string)
        public_key: Authority public key (hex string)
        veto_power: Whether this authority can veto proposals
    """
    id: str
    public_key: str
    veto_power: bool


class GovernanceProtocol:
    """Decentralized governance protocol.
    
    Provides on-chain governance with:
    - Stake-weighted voting
    - Time-locked execution
    - Merkle-logged votes
    - Veto mechanism for emergency stops
    
    Security Properties:
    - All proposals require threshold approval
    - Votes weighted by validator stake (prevents Sybil attacks)
    - Time-locked execution prevents rushed changes
    - All actions auditable via Merkle logs
    """
    
    def __init__(self):
        """Initialize governance protocol."""
        self.proposals: Dict[str, PolicyProposal] = {}
        self.votes: Dict[str, List[Vote]] = {}
        self.proposal_status: Dict[str, ProposalStatus] = {}
        self.executed_proposals: List[str] = []
        self.vetoed_proposals: List[str] = []
        self.current_epoch: int = 0
        self.total_voting_weight: int = 0
        self.authorities: Dict[str, Authority] = {}
    
    def propose_policy_change(self, proposal: PolicyProposal) -> str:
        """Submit a new governance proposal.
        
        Args:
            proposal: Policy proposal to submit
            
        Returns:
            Proposal ID (hex string)
            
        Security:
            - Proposal logged in Merkle tree
            - Audit trail records proposer and timestamp
        """
        proposal_id = proposal.id
        self.proposals[proposal_id] = proposal
        self.votes[proposal_id] = []
        self.proposal_status[proposal_id] = ProposalStatus.ACTIVE
        
        # TODO: Emit audit TXO for proposal submission
        # TODO: Add to Merkle tree
        
        return proposal_id
    
    def vote(self, proposal_id: str, voter: str, vote: Vote) -> bool:
        """Cast a vote on a proposal.
        
        Args:
            proposal_id: Proposal to vote on (hex string)
            voter: Validator ID casting vote (hex string)
            vote: Vote object with decision and weight
            
        Returns:
            True if vote accepted, False otherwise
            
        Security:
            - Vote signature verified
            - Duplicate votes prevented
            - Vote logged in Merkle tree
            - Voting period enforced
        """
        # Check if proposal exists
        if proposal_id not in self.proposals:
            return False
        
        proposal = self.proposals[proposal_id]
        
        # Check if voting is still active
        if self.current_epoch >= proposal.creation_epoch + proposal.voting_period:
            return False
        
        # Check if proposal is still active
        if self.proposal_status[proposal_id] != ProposalStatus.ACTIVE:
            return False
        
        # Check for duplicate votes
        if any(v.voter == voter for v in self.votes[proposal_id]):
            return False
        
        # TODO: Verify vote signature
        
        # Record vote
        self.votes[proposal_id].append(vote)
        
        # TODO: Add vote to Merkle tree
        # TODO: Emit audit TXO for vote
        
        return True
    
    def tally_votes(self, proposal_id: str) -> Dict[str, int]:
        """Calculate vote tally for a proposal.
        
        Args:
            proposal_id: Proposal to tally (hex string)
            
        Returns:
            Dictionary with approve, reject, abstain weights
        """
        if proposal_id not in self.votes:
            return {"approve": 0, "reject": 0, "abstain": 0}
        
        tally = {"approve": 0, "reject": 0, "abstain": 0}
        
        for vote in self.votes[proposal_id]:
            if vote.decision == VoteDecision.APPROVE:
                tally["approve"] += vote.weight
            elif vote.decision == VoteDecision.REJECT:
                tally["reject"] += vote.weight
            elif vote.decision == VoteDecision.ABSTAIN:
                tally["abstain"] += vote.weight
        
        return tally
    
    def is_approved(self, proposal_id: str) -> bool:
        """Check if proposal has reached approval threshold.
        
        Args:
            proposal_id: Proposal to check (hex string)
            
        Returns:
            True if approved, False otherwise
        """
        if proposal_id not in self.proposals:
            return False
        
        proposal = self.proposals[proposal_id]
        tally = self.tally_votes(proposal_id)
        
        if self.total_voting_weight == 0:
            return False
        
        approval_percentage = (tally["approve"] * 100) // self.total_voting_weight
        return approval_percentage >= proposal.threshold
    
    def can_execute(self, proposal_id: str) -> bool:
        """Check if proposal can be executed.
        
        Args:
            proposal_id: Proposal to check (hex string)
            
        Returns:
            True if can be executed, False otherwise
            
        Conditions:
            - Proposal must be approved
            - Voting period must be over
            - Timelock must have elapsed
            - Proposal must not be vetoed or already executed
        """
        if proposal_id not in self.proposals:
            return False
        
        proposal = self.proposals[proposal_id]
        
        # Check if approved
        if not self.is_approved(proposal_id):
            return False
        
        # Check if voting period is over
        if self.current_epoch < proposal.creation_epoch + proposal.voting_period:
            return False
        
        # Check if timelock has elapsed
        if self.current_epoch < proposal.creation_epoch + proposal.voting_period + proposal.timelock:
            return False
        
        # Check if already executed or vetoed
        if proposal_id in self.executed_proposals or proposal_id in self.vetoed_proposals:
            return False
        
        return True
    
    def execute_approved_proposal(self, proposal_id: str) -> bool:
        """Execute an approved proposal.
        
        Args:
            proposal_id: Proposal to execute (hex string)
            
        Returns:
            True if executed successfully, False otherwise
            
        Security:
            - Only approved proposals can execute
            - Timelock must have elapsed
            - Execution logged in Merkle tree
            - Execution is irreversible
        """
        if not self.can_execute(proposal_id):
            return False
        
        # Mark as executed
        self.executed_proposals.append(proposal_id)
        self.proposal_status[proposal_id] = ProposalStatus.EXECUTED
        
        # TODO: Execute proposal payload
        # TODO: Emit audit TXO for execution
        # TODO: Add to Merkle tree
        
        return True
    
    def veto(self, proposal_id: str, authority: Authority) -> bool:
        """Veto a proposal (emergency stop mechanism).
        
        Args:
            proposal_id: Proposal to veto (hex string)
            authority: Authority exercising veto power
            
        Returns:
            True if veto successful, False otherwise
            
        Security:
            - Only authorized entities can veto
            - Cannot veto already executed proposals
            - Veto logged in Merkle tree
        """
        # Check if proposal exists
        if proposal_id not in self.proposals:
            return False
        
        # Check if already executed
        if proposal_id in self.executed_proposals:
            return False
        
        # Check if already vetoed
        if proposal_id in self.vetoed_proposals:
            return False
        
        # Verify authority has veto power
        if not authority.veto_power:
            return False
        
        # TODO: Verify authority signature
        
        # Mark as vetoed
        self.vetoed_proposals.append(proposal_id)
        self.proposal_status[proposal_id] = ProposalStatus.VETOED
        
        # TODO: Emit audit TXO for veto
        # TODO: Add to Merkle tree
        
        return True
    
    def advance_epoch(self):
        """Advance to next epoch."""
        self.current_epoch += 1
    
    def set_total_voting_weight(self, weight: int):
        """Set total voting weight (typically total stake).
        
        Args:
            weight: Total voting weight
        """
        self.total_voting_weight = weight
    
    def register_authority(self, authority: Authority):
        """Register an authority with veto power.
        
        Args:
            authority: Authority to register
        """
        self.authorities[authority.id] = authority


@dataclass(frozen=True)
class GovernanceRule:
    identifier: str
    threshold: float
    weight: float

    def evaluate(self, metrics: dict[str, float]) -> float:
        score = metrics.get(self.identifier, 0.0)
        return self.weight if score >= self.threshold else 0.0


def vote_outcome(votes: dict[str, int]) -> dict[str, object]:
    tally = sum(votes.values()) or 1
    normalized = {k: v / tally for k, v in votes.items()}
    winner = sorted(normalized.items(), key=lambda kv: (-kv[1], kv[0]))[0]
    return {"winner": winner[0], "share": winner[1], "distribution": normalized}


def governance_score(rules: list[GovernanceRule], metrics: dict[str, float]) -> dict[str, float]:
    total_weight = sum(rule.weight for rule in rules) or 1.0
    satisfied = sum(rule.evaluate(metrics) for rule in rules)
    return {"score": satisfied / total_weight, "details": sum(metrics.values())}


def deterministic_auction(bids: dict[str, float]) -> dict[str, object]:
    ordered = sorted(bids.items(), key=lambda kv: (-kv[1], kv[0]))
    winner, price = ordered[0]
    clearing_price = ordered[1][1] if len(ordered) > 1 else price
    return {"winner": winner, "clearing_price": clearing_price, "ordered": ordered}

File: qratum-rust/src/governance.rs
//! # Governance Module - Rust Protocol Layer
//!
//! ## Lifecycle Stage: Governance Execution
//!
//! This module provides Rust-side governance structures that integrate with
//! the Python governance protocol for decentralized protocol changes.
//!
//! ## Architectural Role
//!
//! - **Proposal Types**: Define governance proposal structures
//! - **Voting Mechanisms**: Track votes and calculate outcomes
//! - **Execution Hooks**: Execute approved governance proposals
//! - **Stake-Weighted Voting**: Weight votes by validator stake
//!
//! ## Security Rationale
//!
//! - All proposals require threshold approval
//! - Votes are weighted by stake (prevents Sybil attacks)
//! - Time-locked execution prevents rushed changes
//! - Veto mechanism for emergency stops
//!
//! ## Audit Trail
//!
//! - All proposals logged with proposer and details
//! - All votes recorded with voter and weight
//! - Execution events logged with timestamp
//! - Veto events recorded with authority


extern crate alloc;
use alloc::collections::BTreeMap;
use alloc::vec::Vec;
use alloc::string::String;

/// Proposal identifier
pub type ProposalID = [u8; 32];

/// Voter identifier (typically validator ID)
pub type VoterID = [u8; 32];

/// Authority identifier (for veto power)
pub type AuthorityID = [u8; 32];

/// Governance proposal type
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum ProposalType {
    /// Change protocol parameter
    ParameterChange,
    /// Protocol upgrade
    ProtocolUpgrade,
    /// Validator set change
    ValidatorSetChange,
    /// Treasury spending
    TreasurySpending,
    /// Emergency action
    Emergency,
}

/// Vote decision
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum VoteDecision {
    /// Vote in favor
    Approve,
    /// Vote against
    Reject,
    /// Abstain from voting
    Abstain,
}

/// Governance vote
#[derive(Debug, Clone)]
pub struct GovernanceVote {
    /// Voter identifier
    pub voter: VoterID,
    
    /// Vote decision
    pub decision: VoteDecision,
    
    /// Voting weight (typically stake amount)
    pub weight: u64,
    
    /// Vote signature
    pub signature: [u8; 64],
    
    /// Epoch when vote was cast
    pub epoch: u64,
}

/// Governance proposal
#[derive(Debug, Clone)]
pub struct GovernanceProposal {
    /// Unique proposal identifier
    pub id: ProposalID,
    
    /// Proposal type
    pub proposal_type: ProposalType,
    
    /// Proposer identifier
    pub proposer: VoterID,
    
    /// Proposal description
    pub description: String,
    
    /// Proposal payload (implementation-specific)
    pub payload: Vec<u8>,
    
    /// Required approval threshold (percentage, 0-100)
    pub threshold: u8,
    
    /// Voting period (number of epochs)
    pub voting_period: u64,
    
    /// Timelock period after approval (epochs before execution)
    pub timelock: u64,
    
    /// Epoch when proposal was created
    pub creation_epoch: u64,
}

impl GovernanceProposal {
    /// Check if voting period is active
    pub fn is_voting_active(&self, current_epoch: u64) -> bool {
        current_epoch < self.creation_epoch + self.voting_period
    }
    
    /// Check if proposal is approved (based on vote tallies)
    pub fn is_approved(&self, approve_weight: u64, total_weight: u64) -> bool {
        if total_weight == 0 {
            return false;
        }
        
        let approval_percentage = (approve_weight * 100) / total_weight;
        approval_percentage >= self.threshold as u64
    }
    
    /// Check if proposal can be executed
    pub fn can_execute(&self, current_epoch: u64, is_approved: bool) -> bool {
        // Must be approved
        if !is_approved {
            return false;
        }
        
        // Voting period must be over
        if current_epoch < self.creation_epoch + self.voting_period {
            return false;
        }
        
        // Timelock must have elapsed
        if current_epoch < self.creation_epoch + self.voting_period + self.timelock {
            return false;
        }
        
        true
    }
}

/// Governance state
///
/// ## Security Invariants
/// - Only approved proposals can be executed
/// - Proposals respect voting period and timelock
/// - Vetoed proposals cannot be executed
/// - All votes are stake-weighted
pub struct GovernanceState {
    /// Active proposals
    pub proposals: BTreeMap<ProposalID, GovernanceProposal>,
    
    /// Votes for each proposal
    pub votes: BTreeMap<ProposalID, Vec<GovernanceVote>>,
    
    /// Executed proposals
    pub executed: Vec<ProposalID>,
    
    /// Vetoed proposals
    pub vetoed: Vec<ProposalID>,
    
    /// Current epoch
    pub current_epoch: u64,
    
    /// Total voting weight (typically total stake)
    pub total_voting_weight: u64,
}

impl GovernanceState {
    /// Create new governance state
    pub fn new() -> Self {
        Self {
            proposals: BTreeMap::new(),
            votes: BTreeMap::new(),
            executed: Vec::new(),
            vetoed: Vec::new(),
            current_epoch: 0,
            total_voting_weight: 0,
        }
    }
    
    /// Submit a new governance proposal
    pub fn submit_proposal(&mut self, proposal: GovernanceProposal) {
        let id = proposal.id;
        self.proposals.insert(id, proposal);
        self.votes.insert(id, Vec::new());
        
        // TODO: Emit audit TXO for proposal submission
    }
    
    /// Cast a vote on a proposal
    pub fn vote(&mut self, proposal_id: ProposalID, vote: GovernanceVote) -> bool {
        // Check if proposal exists
        let proposal = match self.proposals.get(&proposal_id) {
            Some(p) => p,
            None => return false,
        };
        
        // Check if voting is still active
        if !proposal.is_voting_active(self.current_epoch) {
            return false;
        }
        
        // Check for duplicate votes
        if let Some(votes) = self.votes.get(&proposal_id) {
            if votes.iter().any(|v| v.voter == vote.voter) {
                return false; // Already voted
            }
        }
        
        // Add vote
        self.votes.get_mut(&proposal_id).unwrap().push(vote);
        
        // TODO: Emit audit TXO for vote
        
        true
    }
    
    /// Calculate vote tally for a proposal
    pub fn tally_votes(&self, proposal_id: &ProposalID) -> (u64, u64, u64) {
        let votes = match self.votes.get(proposal_id) {
            Some(v) => v,
            None => return (0, 0, 0),
        };
        
        let mut approve = 0u64;
        let mut reject = 0u64;
        let mut abstain = 0u64;
        
        for vote in votes {
            match vote.decision {
                VoteDecision::Approve => approve += vote.weight,
                VoteDecision::Reject => reject += vote.weight,
                VoteDecision::Abstain => abstain += vote.weight,
            }
        }
        
        (approve, reject, abstain)
    }
    
    /// Execute an approved proposal
    pub fn execute_proposal(&mut self, proposal_id: ProposalID) -> bool {
        // Check if proposal exists
        let proposal = match self.proposals.get(&proposal_id) {
            Some(p) => p.clone(),
            None => return false,
        };
        
        // Check if already executed
        if self.executed.contains(&proposal_id) {
            return false;
        }
        
        // Check if vetoed
        if self.vetoed.contains(&proposal_id) {
            return false;
        }
        
        // Calculate approval
        let (approve, _, _) = self.tally_votes(&proposal_id);
        let is_approved = proposal.is_approved(approve, self.total_voting_weight);
        
        // Check if can execute
        if !proposal.can_execute(self.current_epoch, is_approved) {
            return false;
        }
        
        // Execute proposal (implementation-specific)
        // TODO: Dispatch to appropriate handler based on proposal_type
        
        // Mark as executed
        self.executed.push(proposal_id);
        
        // TODO: Emit audit TXO for execution
        
        true
    }
    
    /// Veto a proposal
    pub fn veto(&mut self, proposal_id: ProposalID, _authority: AuthorityID) -> bool {
        // Check if proposal exists
        if !self.proposals.contains_key(&proposal_id) {
            return false;
        }
        
        // Check if already executed
        if self.executed.contains(&proposal_id) {
            return false; // Cannot veto executed proposals
        }
        
        // Check if already vetoed
        if self.vetoed.contains(&proposal_id) {
            return false;
        }
        
        // TODO: Verify authority has veto power
        
        // Mark as vetoed
        self.vetoed.push(proposal_id);
        
        // TODO: Emit audit TXO for veto
        
        true
    }
    
    /// Advance to next epoch
    pub fn advance_epoch(&mut self) {
        self.current_epoch += 1;
    }
    
    /// Get proposal by ID
    pub fn get_proposal(&self, id: &ProposalID) -> Option<&GovernanceProposal> {
        self.proposals.get(id)
    }
}

impl Default for GovernanceState {
    fn default() -> Self {
        Self::new()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use alloc::vec;
    
    #[test]
    fn test_governance_proposal() {
        let proposal = GovernanceProposal {
            id: [1u8; 32],
            proposal_type: ProposalType::ParameterChange,
            proposer: [2u8; 32],
            description: "Test proposal".into(),
            payload: vec![],
            threshold: 67, // 67% approval required
            voting_period: 10,
            timelock: 5,
            creation_epoch: 0,
        };
        
        assert!(proposal.is_voting_active(5));
        assert!(!proposal.is_voting_active(11));
    }
    
    #[test]
    fn test_governance_state() {
        let mut state = GovernanceState::new();
        state.total_voting_weight = 1000;
        
        let proposal = GovernanceProposal {
            id: [1u8; 32],
            proposal_type: ProposalType::ParameterChange,
            proposer: [2u8; 32],
            description: "Test proposal".into(),
            payload: vec![],
            threshold: 67,
            voting_period: 10,
            timelock: 5,
            creation_epoch: 0,
        };
        
        state.submit_proposal(proposal);
        
        // Vote
        let vote = GovernanceVote {
            voter: [3u8; 32],
            decision: VoteDecision::Approve,
            weight: 700,
            signature: [0u8; 64],
            epoch: 0,
        };
        
        let voted = state.vote([1u8; 32], vote);
        assert!(voted);
        
        // Check tally
        let (approve, reject, abstain) = state.tally_votes(&[1u8; 32]);
        assert_eq!(approve, 700);
        assert_eq!(reject, 0);
        assert_eq!(abstain, 0);
    }
}
